File: scripts/parse_thermo.mjs
import fs from 'node:fs';
import path from 'node:path';

const mdPath = "C:\\Users\\LENOVO\\Downloads\\Thermodynamics_All_Questions_Pooled_Topicwise.md";
const outPath = "D:\\New folder\\neetcore.com\\src\\data\\thermodynamics-questions.ts";

const userSubtopics = [
    "Thermodynamic Basics: Systems, Properties & Processes",
    "First Law of Thermodynamics: Heat, Work & Internal Energy",
    "Enthalpy and ΔH–ΔU Relations",
    "Work Done in Thermodynamic Processes",
    "Entropy and Second/Third Law of Thermodynamics",
    "Gibbs Free Energy, Spontaneity & Equilibrium",
    "Thermochemistry: Exothermic/Endothermic Reactions",
    "Kirchhoff’s Equation & Temperature Dependence of ΔH",
    "Enthalpy of Formation",
    "Enthalpy of Combustion",
    "Enthalpy of Hydrogenation, Atomisation & Transition",
    "Bond Energy / Bond Enthalpy",
    "Hess’s Law and Indirect Enthalpy Calculations",
    "Heat Capacity and the Cp–Cv Relation",
];

const text = fs.readFileSync(mdPath, 'utf-8');

const sections = text.split(/\n##\s+/);
const questionSections = [];
for (const sec of sections) {
  if (sec.includes("### Question")) {
    const title = sec.split("\n")[0].trim();
    questionSections.push([title, sec]);
  }
}
console.log(`Found ${questionSections.length} sections`);
questionSections.forEach(([t],i)=> console.log(i+1, t.slice(0,60)));

function addMath(s){
  if(!s) return s;
  s = s.replace(/\s+/g,' ').trim();
  // Replace symbols
  s = s.replace(/∆/g, '\\Delta');
  s = s.replace(/Δ/g, '\\Delta');
  s = s.replace(/→/g, '\\rightarrow');
  s = s.replace(/→/g, '\\rightarrow');
  s = s.replace(/   /g, '\\rightleftharpoons');
  s = s.replace(/   /g, '\\rightleftharpoons');
  s = s.replace(/⇌/g, '\\rightleftharpoons');
  s = s.replace(/½/g, '\\frac{1}{2}');
  s = s.replace(/¼/g, '\\frac{1}{4}');
  s = s.replace(/¾/g, '\\frac{3}{4}');
  // Wrap Delta, arrows, fractions - single pass each
  s = s.replace(/\\Delta\s*([A-Za-z_]+)?/g, (m, g1)=> g1 ? `$\\Delta ${g1}$` : `$\\Delta$`);
  s = s.replace(/\\rightarrow/g, '$\\rightarrow$');
  s = s.replace(/\\rightleftharpoons/g, '$\\rightleftharpoons$');
  // Wrap fractions
  s = s.replace(/\\frac\{[^}]+\}\{[^}]+\}/g, (m)=> `$${m}$`);
  // Fix adjacent math segments that now have "$$" between them - insert space instead of collapsing
  s = s.replace(/\$\$/g, '$ $');
  // Clean up any "$ $" double spaces
  s = s.replace(/\$ \$/g, '$ $');
  s = s.replace(/\bCp\b/g, '$C_p$');
  s = s.replace(/\bCv\b/g, '$C_v$');
  s = s.replace(/\$\\Delta n_g\$/g, '$\\Delta n_g$');
  if(s.endsWith(" N")) s = s.slice(0,-2).trim();
  return s.trim();
}
function cleanOpt(opt){
  opt = opt.trim();
  // Only strip leading dash if it's a separator " - " not a negative sign
  // Negative is "--" or "-500" without space; separator is "- " with space
  if(opt.startsWith("- ") ) opt = opt.slice(2).trim();
  opt = addMath(opt);
  return opt;
}

const imageMap = {
  "110944.png": "/images/che/thermo/q56-reversible-expansion.png",
  "111003.png": "/images/che/thermo/q57-pv-plots.png",
  "01_06_01 PM.png": "/images/che/thermo/q58-pv-curve.png",
  "111102.png": "/images/che/thermo/q82-a-to-b-path.png",
};

let all = [];
let globalIdx=0;
for(let secIdx=0; secIdx<questionSections.length; secIdx++){
  const [title, sec] = questionSections[secIdx];
  const canon = userSubtopics[secIdx] || title;
  const qBlocks = sec.split(/###\s+Question\s+\d+/);
  const qs=[];
  for(let b=1;b<qBlocks.length;b++){
    const block = qBlocks[b];
    globalIdx++;
    let qMatch = block.match(/\*\*Question\*\*\s*([\s\S]*?)\s*\*\*Options\*\*/);
    let qText = qMatch ? qMatch[1].trim() : "";
    let imgPublic = null;
    const imgCands = [...block.matchAll(/[A-Z]:\\[^\n"]+\.png/g)].map(m=>m[0]);
    for(const cand of imgCands){
      for(const [k,pub] of Object.entries(imageMap)){
        if(cand.includes(k)){ imgPublic = pub; break; }
      }
      if(!imgPublic){
        const fname = path.basename(cand).replace(/ /g,"_").replace(/,/g,"");
        imgPublic = `/images/che/thermo/${fname}`;
      }
      qText = qText.replace(cand,'').replace(`"${cand}"`,'');
    }
    qText = qText.replace(/"C:\\[^"]+\.png"/g,'').replace(/'C:\\[^']+\.png'/g,'').replace(/C:\\[^\s]+\.png/g,'').trim().replace(/^"+|"+$/g,'');
    qText = qText.replace(/\s+/g,' ').trim();

    let oMatch = block.match(/\*\*Options\*\*\s*([\s\S]*?)\s*\*\*Correct Option/);
    let oText = oMatch ? oMatch[1].trim() : "";
    oText = oText.replace(/[A-Z]:\\[^\n"]+\.png/g,'').replace(/"C:\\[^"]+\.png"/g,'');

    // Parse options
    const optRegex = /\(\s*([1-4])\s*\)\s*([\s\S]*?)(?=(?:-\s*\(\s*[1-4]\s*\)|$))/g;
    // Simpler: find all (n) segments by locating positions of " - (n)" or "(n)"
    // Let's use manual split approach: find all occurrences of "(1)" etc.
    let rawOpts = [];
    const optIter = [...oText.matchAll(/\(\s*([1-4])\s*\)\s*([\s\S]*?)(?=\s*-\s*\(\s*[1-4]\s*\)|\s*\*\*Correct|\s*$)/g)];
    // Fallback if regex above fails
    if(optIter.length!==4){
      // Try split by "- ("
      const parts = oText.split(/\s*-\s*\(\s*[1-4]\s*\)\s*/);
      // parts[0] may be empty or content before first?
      // The pattern "- (1) content - (2) content" => split gives ["", "content ", "content " ...]
      const filtered = parts.filter(p=>p.trim().length>0).map(p=> p.replace(/^--/,'-').trim());
       if(filtered.length===4){
         for(let i=0;i<4;i++) rawOpts.push([String(i+1), filtered[i]]);
       } else {
         console.log(`Warning Q${globalIdx} opts len ${optIter.length} oText: ${oText.slice(0,150)}`);
         // try another method: find by scanning
         const matches = [...oText.matchAll(/\(\s*([1-4])\s*\)/g)];
         for(let i=0;i<matches.length;i++){
           const start = matches[i].index + matches[i][0].length;
           const end = i+1<matches.length ? matches[i+1].index : oText.length;
           const content = oText.slice(start,end).replace(/^\s*-\s*/,'').replace(/^--/,'-').trim();
           rawOpts.push([matches[i][1], content]);
         }
      }
    } else {
      rawOpts = optIter.map(m=>[m[1], m[2]]);
    }
    if(rawOpts.length!==4){
      // last fallback: use optIter
      rawOpts = optIter.map(m=>[m[1], m[2]]);
    }
    let options=[];
    for(const [num, content] of rawOpts){
      let c = content.trim().replace(/^--/,'-').trim();
      // Do not strip single leading dash that is negative sign; only strip separator artefact " - " already removed
      c = c.replace(/\s+/g,' ').trim();
      // Convert remaining "--" inside to "-" (preserve negative)
      c = c.replace(/--/g,'-');
      options.push(c);
    }
    while(options.length<4) options.push("");
    options = options.slice(0,4);

    const cMatch = block.match(/\*\*Correct Option:\*\*\s*\(\s*([1-4])\s*\)/);
    const correctNum = cMatch ? parseInt(cMatch[1]) : 1;
    const answer = {1:'A',2:'B',3:'C',4:'D'}[correctNum];

    const qFormatted = addMath(qText);
    const letters=['A','B','C','D'];
    const formattedOpts = options.map((opt,i)=>{
      const clean = cleanOpt(opt);
      if(!clean.startsWith("(")) return `(${letters[i]}) ${clean}`;
      return clean;
    });

    qs.push({q: qFormatted, options: formattedOpts, answer, imagePublic: imgPublic});
  }
  all.push([canon, qs]);
}

for(const [name, qs] of all) console.log(`${name}: ${qs.length}`);
console.log(`Total ${all.reduce((a,[_,q])=>a+q.length,0)}`);

let out = [];
out.push("import type { SetData } from './questions';");
out.push("");
out.push("export const thermodynamicsNcertEssentials: SetData = {");
out.push("  subtopics: [");
for(const [name, qs] of all){
  out.push(`    {`);
  out.push(`      name: ${JSON.stringify(name)},`);
  out.push(`      questions: [`);
  for(const qobj of qs){
    const qEsc = JSON.stringify(qobj.q);
    const optsEsc = JSON.stringify(qobj.options);
    const ansEsc = JSON.stringify(qobj.answer);
    if(qobj.imagePublic){
      const imgEsc = JSON.stringify(qobj.imagePublic);
      out.push(`        { q: ${qEsc}, options: ${optsEsc} as [string, string, string, string], answer: ${ansEsc} as const, imagePublic: ${imgEsc} },`);
    } else {
      out.push(`        { q: ${qEsc}, options: ${optsEsc} as [string, string, string, string], answer: ${ansEsc} as const },`);
    }
  }
  out.push(`      ],`);
  out.push(`    },`);
}
out.push("  ],");
out.push("};");
out.push("");

fs.writeFileSync(outPath, out.join("\n"), 'utf-8');
console.log(`Wrote to ${outPath}`);

File: scripts/parse_thermo.py
import re
import pathlib
import json

md_path = pathlib.Path(r"C:\Users\LENOVO\Downloads\Thermodynamics_All_Questions_Pooled_Topicwise.md")
out_path = pathlib.Path(r"D:\New folder\neetcore.com\src\data\thermodynamics-questions.ts")

user_subtopics = [
    "Thermodynamic Basics: Systems, Properties & Processes",
    "First Law of Thermodynamics: Heat, Work & Internal Energy",
    "Enthalpy and ΔH–ΔU Relations",
    "Work Done in Thermodynamic Processes",
    "Entropy and Second/Third Law of Thermodynamics",
    "Gibbs Free Energy, Spontaneity & Equilibrium",
    "Thermochemistry: Exothermic/Endothermic Reactions",
    "Kirchhoff’s Equation & Temperature Dependence of ΔH",
    "Enthalpy of Formation",
    "Enthalpy of Combustion",
    "Enthalpy of Hydrogenation, Atomisation & Transition",
    "Bond Energy / Bond Enthalpy",
    "Hess’s Law and Indirect Enthalpy Calculations",
    "Heat Capacity and the Cp–Cv Relation",
]

# Map markdown heading normalized -> user subtopic
# markdown headings: e.g., "Thermodynamic Basics: Systems, Properties & Processes"
# We'll just use order, but also build map for safety
md_to_user = {}
# manual mapping - assume order same
text = md_path.read_text(encoding="utf-8")

# Find all ## sections
# Pattern: ## <title>\n ... until next ##
sections = re.split(r'\n##\s+', text)
# First section is title + intro, ignore
# sections[0] is header before first ##, sections[1..] are subtopics + questions
# But the initial subtopics list is also "## Subtopics created" which is not a real section
# Real question sections start at "## Thermodynamic Basics"

# Filter to keep only those that contain "### Question"
question_sections = []
for sec in sections:
    if "### Question" in sec:
        # title is first line
        title = sec.split("\n")[0].strip()
        question_sections.append((title, sec))

print(f"Found {len(question_sections)} sections with questions")

# Should be 14
for i, (title, sec) in enumerate(question_sections):
    print(i+1, title[:60])

# Define helper to clean and add LaTeX
def add_math(s: str) -> str:
    if not s:
        return s
    # Normalize whitespace
    s = re.sub(r'\s+', ' ', s).strip()
    # Remove some artifact markers like "N" at end (from PDF extraction)
    # Already handled elsewhere
    # Replace Δ/∆ with LaTeX
    # We will wrap ΔX directly as $\\Delta X$
    # Replace occurrences of ∆ or Δ followed by letters
    # Use placeholder: replace ∆H, ΔH, etc.
    # General: ∆ or Δ + optional space + letters (H, S, G, E, U, Cp, Cv, ng)
    # We'll do simple replacements first
    replacements = [
        (r'∆', r'\\Delta'),
        (r'Δ', r'\\Delta'),
        (r'→', r'\\rightarrow'),
        (r'→', r'\\rightarrow'),
        (r'   ', r'\\rightleftharpoons'),
        (r'   ', r'\\rightleftharpoons'),
        (r'⇌', r'\\rightleftharpoons'),
        (r'½', r'\\frac{1}{2}'),
        (r'¼', r'\\frac{1}{4}'),
        (r'¾', r'\\frac{3}{4}'),
        (r'–', r'-'), # en dash to hyphen for now, but later for ΔH–ΔU keep hyphen
        (r'—', r'-'),
    ]
    for pat, rep in replacements:
        s = s.replace(pat, rep)
    # Now wrap \Delta occurrences with $...$
    # Pattern: \Delta followed by optional space and letters/numbers
    # We want to turn "\Delta H" or "\DeltaH" into "$\\Delta H$"
    # But after replacement, we have "\Delta" literal
    s = re.sub(r'\\Delta\s*([A-Za-z_]+)', r'$\\Delta \1$', s)
    s = re.sub(r'\\Delta', r'$\\Delta$', s) # standalone
    # Wrap \rightarrow with $...$
    s = s.replace(r'\rightarrow', r'$\rightarrow$')
    s = s.replace(r'\rightleftharpoons', r'$\rightleftharpoons$')
    # Wrap fractions: \frac is already math, need $...$
    s = re.sub(r'\\frac\{[^}]+\}\{[^}]+\}', lambda m: f'${m.group(0)}$', s)
    # Handle Cp, Cv -> $C_p$ etc.
    s = re.sub(r'\bCp\b', r'$C_p$', s)
    s = re.sub(r'\bCv\b', r'$C_v$', s)
    s = re.sub(r'\bCp--Cv\b', r'$C_p - C_v$', s)
    s = re.sub(r'\bCp–Cv\b', r'$C_p - C_v$', s)
    # Also handle \Delta ng -> \Delta n_g
    s = s.replace(r'$\Delta ng$', r'$\Delta n_g$')
    s = s.replace(r'$\Delta ng$', r'$\Delta n_g$')
    # Clean double math wrappers like $$...$$ from previous
    s = re.sub(r'\$\$+', '$', s)
    # Remove stray trailing " N" artifacts that are at end of question (from PDF)
    # e.g., question ends with " N"
    if s.endswith(" N"):
        s = s[:-2].strip()
    # Fix duplicate $ issues: "$\Delta$" should be okay, but "$\Delta H$" repeated?
    # Ensure we don't have $ inside already $...$
    # Simplify: collapse "$ $" -> ""
    # Remove empty?
    # Keep as is
    return s.strip()

def clean_option_text(opt: str) -> str:
    # Remove leading numbering handled outside, just clean math
    opt = opt.strip()
    # Remove leading dash if any
    if opt.startswith("-"):
        opt = opt[1:].strip()
    # Apply math formatting
    opt = add_math(opt)
    # Ensure options have parentheses style? Keep as is
    return opt

# Image mapping
image_map = {
    "110944.png": "/images/che/thermo/q56-reversible-expansion.png",
    "111003.png": "/images/che/thermo/q57-pv-plots.png",
    "01_06_01 PM.png": "/images/che/thermo/q58-pv-curve.png",
    "111102.png": "/images/che/thermo/q82-a-to-b-path.png",
}
# Also handle the ChatGPT Image with commas
# We'll map by substring

# Parse questions
all_questions_by_section = []
global_q_index = 0

for sec_idx, (title, sec) in enumerate(question_sections):
    # Determine user subtopic name - map by index order (should match)
    if sec_idx < len(user_subtopics):
        canon_name = user_subtopics[sec_idx]
    else:
        canon_name = title
    # Split by ### Question
    q_blocks = re.split(r'###\s+Question\s+\d+', sec)
    # First block is heading, ignore
    qs = []
    for block in q_blocks[1:]:
        global_q_index += 1
        # Extract question text between **Question** and **Options**
        # Use regex
        q_match = re.search(r'\*\*Question\*\*\s*(.*?)\s*\*\*Options\*\*', block, re.DOTALL)
        q_text = q_match.group(1).strip() if q_match else ""
        # Remove image paths from q_text
        img_public = None
        img_candidates = re.findall(r'[A-Z]:\\[^\n"]+\.png', block)
        for cand in img_candidates:
            # Determine public path
            for key, pub in image_map.items():
                if key in cand:
                    img_public = pub
                    break
            if not img_public:
                # fallback generic
                fname = pathlib.Path(cand).name.replace(" ", "_").replace(",", "")
                img_public = f"/images/che/thermo/{fname}"
            # Remove this path from q_text
            q_text = q_text.replace(cand, "")
            q_text = q_text.replace(f'"{cand}"', "")
            q_text = q_text.replace(cand.replace("\\", "/"), "")
        # Also handle quoted path with quotes
        q_text = re.sub(r'"C:\\[^"]+\.png"', '', q_text).strip()
        q_text = re.sub(r"'C:\\[^']+\.png'", '', q_text).strip()
        q_text = re.sub(r'C:\\[^\s]+\.png', '', q_text).strip()
        # Clean up stray quotes
        q_text = q_text.strip().strip('"').strip("'")
        q_text = re.sub(r'\s+', ' ', q_text).strip()
        # Extract options block between **Options** and **Correct Option**
        o_match = re.search(r'\*\*Options\*\*\s*(.*?)\s*\*\*Correct Option', block, re.DOTALL)
        o_text = o_match.group(1).strip() if o_match else ""
        # Remove image path from o_text if any (unlikely)
        o_text = re.sub(r'[A-Z]:\\[^\n"]+\.png', '', o_text)
        o_text = re.sub(r'"C:\\[^"]+\.png"', '', o_text)
        # Now split options into 4
        # Options are like "- (1) ... - (2) ... - (3) ... - (4) ..."
        # Use regex to find each (1)..(4)
        opt_pattern = re.compile(r'\(\s*([1-4])\s*\)\s*(.*?)(?=(?:-\s*\(\s*[1-4]\s*\)|\Z))', re.DOTALL)
        raw_opts = opt_pattern.findall(o_text)
        # If not 4, try alternative split by " - "
        if len(raw_opts) != 4:
            # Try split by pattern "- ("
            parts = re.split(r'-\s*\(\s*[1-4]\s*\)', o_text)
            # parts[0] is empty before first, then 4 parts?
            # This fallback may produce 4
            raw_opts2 = []
            # Use findall again with simpler
            for m in re.finditer(r'\(\s*[1-4]\s*\)', o_text):
                pass
            print(f"Warning Q{global_q_index} opts len {len(raw_opts)} o_text: {o_text[:120]}")
            # Try to extract manually
            # Split by "- ("
            splits = re.split(r'\s*-\s*\(\s*[1-4]\s*\)\s*', o_text)
            # splits will have 5 elements (first empty)
            # The first split before first option is empty or dash
            # Clean
            splits = [s for s in splits if s.strip()]
            # If we have 4, use them
            if len(splits) == 4:
                raw_opts = [(str(i+1), splits[i]) for i in range(4)]
        # Ensure 4
        options = []
        for num, content in raw_opts:
            content = content.strip()
            # Remove leading "-"
            content = content.lstrip("-").strip()
            # Remove trailing hyphens/separators like "------------------------------------------------------------------------"
            content = re.sub(r'-{2,}', '', content).strip()
            # Remove trailing " Re-NEET (UG) 2022" etc? Keep as is but clean
            options.append(content)
        # If still not 4, pad
        while len(options) < 4:
            options.append("")
        options = options[:4]
        # Map correct
        c_match = re.search(r'\*\*Correct Option:\*\*\s*\(\s*([1-4])\s*\)', block)
        correct_num = int(c_match.group(1)) if c_match else 1
        answer_letter = {1:'A',2:'B',3:'C',4:'D'}[correct_num]
        # Format question and options with math
        q_formatted = add_math(q_text)
        # If q_text already contains chemical equation, ensure it's not over-wrapped
        # Add prefix? Keep as is
        # For options, format each and add (A) prefix for display consistency
        formatted_opts = []
        letters = ['A','B','C','D']
        for i, opt in enumerate(options):
            opt_clean = clean_option_text(opt)
            # Add letter prefix like "(A) " if not already
            if not opt_clean.startswith("("):
                opt_clean = f"({letters[i]}) {opt_clean}"
            else:
                # Ensure letter matches (convert (1) to (A))
                # Already handled? Options content doesn't include numbering after split, so we add prefix
                pass
            # If opt already has letter, don't duplicate
            formatted_opts.append(opt_clean)
        # Ensure options length 4 and are tuple
        qs.append({
            "q": q_formatted,
            "options": formatted_opts,
            "answer": answer_letter,
            "imagePublic": img_public
        })
    all_questions_by_section.append((canon_name, qs))

# Verify counts
for name, qs in all_questions_by_section:
    print(f"{name}: {len(qs)}")

total = sum(len(qs) for _, qs in all_questions_by_section)
print(f"Total {total}")

# Generate TS file
import textwrap

output = []
output.append("import type { SetData } from './questions';")
output.append("")
output.append("export const thermodynamicsNcertEssentials: SetData = {")
output.append("  subtopics: [")

for name, qs in all_questions_by_section:
    # Escape name for TS string: use json dumps for name
    output.append(f"    {{")
    output.append(f"      name: {json.dumps(name, ensure_ascii=False)},")
    output.append(f"      questions: [")
    for qobj in qs:
        q_escaped = json.dumps(qobj['q'], ensure_ascii=False)
        opts_escaped = json.dumps(qobj['options'], ensure_ascii=False)
        ans_escaped = json.dumps(qobj['answer'])
        if qobj['imagePublic']:
            img_escaped = json.dumps(qobj['imagePublic'])
            output.append(f"        {{ q: {q_escaped}, options: {opts_escaped} as [string, string, string, string], answer: {ans_escaped} as const, imagePublic: {img_escaped} }},")
        else:
            output.append(f"        {{ q: {q_escaped}, options: {opts_escaped} as [string, string, string, string], answer: {ans_escaped} as const }},")
    output.append(f"      ],")
    output.append(f"    }},")

output.append("  ],")
output.append("};")
output.append("")

out_path.write_text("\n".join(output), encoding="utf-8")
print(f"Wrote to {out_path}")
